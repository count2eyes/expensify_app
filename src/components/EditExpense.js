import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  onRemove = () => {
    // earlier code
    // this.props.removeExpense({ id: this.props.expense.id });
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(el => el.id === props.match.params.id)
  };
};

// also passed 'props' as 2nd argument for removeExpense required props
//
// const mapDispatchToProps = (dispatch, props) => ({
//   editExpense: (id, expense) => dispatch(editExpense(id, expense)),
//   removeExpense: id => dispatch(removeExpense(id))
// });

const mapDispatchToProps = dispatch => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: id => dispatch(removeExpense({ id }))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
