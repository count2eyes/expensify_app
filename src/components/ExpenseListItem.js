import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";

const ExpenseListItem = ({ description, amount, createdAt, id }) => {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>{" "}
      </Link>
      <p>
        ${amount / 100} - {moment(createdAt).format("ddd DD MMM, YYYY")}
      </p>
    </div>
  );
};

export default ExpenseListItem;
