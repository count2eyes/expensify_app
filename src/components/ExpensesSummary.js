import React from "react";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expenses-total";
import { connect } from "react-redux";

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseString = expenseCount > 1 ? "expenses" : "expense";
  const formattedExpensesTotal = numeral(expensesTotal / 100).format("$0,0.00");

  return (
    <div>
      <h4>
        Viewing {expenseCount} {expenseString} totalling{" "}
        {formattedExpensesTotal}
      </h4>
    </div>
  );
};
const mapStateToProps = state => {
  const visilbleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visilbleExpenses.length,
    expensesTotal: selectExpensesTotal(visilbleExpenses)
  };
};
export default connect(mapStateToProps)(ExpensesSummary);
