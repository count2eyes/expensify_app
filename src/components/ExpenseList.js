import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-wide-screens">Expense</div>
      <div className="show-for-wide-screens">Amount</div>
    </div>

    <div className="list-body">
      {props.expenses.length === 0 ? (
        <div className="list-item list-item--message">
          <span>
            <h3>No expenses</h3>
          </span>
        </div>
      ) : (
        props.expenses.map(el => {
          return <ExpenseListItem {...el} key={el.id} />;
        })
      )}
    </div>
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
