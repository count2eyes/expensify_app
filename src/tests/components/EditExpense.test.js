import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpense";
import expenses from "../fixtures/expenses";

let history, editExpense, removeExpense, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = {
    push: jest.fn()
  };
  wrapper = shallow(
    <EditExpensePage
      history={history}
      editExpense={editExpense} //2args
      removeExpense={removeExpense}
      expense={expenses[0]}
    />
  );
});

test("should render EditExpensePage", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle EditExpense", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test("should handle removeExpense", () => {
  wrapper.find("button").simulate("click");
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id);
});
