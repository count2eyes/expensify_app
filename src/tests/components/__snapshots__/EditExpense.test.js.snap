// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 2650,
        "createdAt": 0,
        "description": "coffee",
        "id": 1,
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
